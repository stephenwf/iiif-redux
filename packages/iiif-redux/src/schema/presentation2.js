/**
 * Presentation 2 - schema definition
 * ============================================================================
 *
 * - Collection
 * - Manifest
 * - Sequence
 * - Canvas
 * - Annotation
 * - AnnotationList
 * - Range
 * - Layer
 * - Image Content (Annotation)
 * - Other Content (Annotation List)
 *
 * - External resources
 * - Services
 */

// @todo Image Service API.
import { compose } from 'redux';
import { schema, normalize } from 'normalizr';
import moveStartCanvasToSequence from '../compat/moveStartCanvasToSequence';
import preprocessLinkedEntities from '../compat/preprocessLinkedEntities';
import addMissingIds from '../compat/addMissingIds';

function createEntity(name) {
  const options = {
    idAttribute: '@id',
  };

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections');
const manifest = createEntity('manifests');
const sequence = createEntity('sequences');
const canvas = createEntity('canvases');
const annotation = createEntity('annotations');
const annotationList = createEntity('annotationLists');
const range = createEntity('ranges');
const layer = createEntity('layers');
const imageResource = createEntity('imageResources'); // thumbnails + image service
const choice = createEntity('choices');

// Unofficial types?
const externalResource = createEntity('externalResources');
const service = createEntity('services', true);
const within = new schema.Array(
  {
    collection,
    manifest,
    externalResource,
    layer,
  },
  entity => {
    switch (entity['@type']) {
      case 'sc:Collection':
        return 'collection';
      case 'sc:Manifest':
        return 'manifest';
      case 'sc:Layer':
        return 'layer';
      default:
        return 'externalResource';
    }
  }
);

const rangeMember = new schema.Union(
  {
    canvas,
    range,
  },
  input => (input['@type'] === 'sc:Canvas' ? 'canvas' : 'range')
);

/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - First Sequence (Optional)
 * - Nth Sequence (Required)
 * - Canvas (Optional)
 * - Annotation (Recommended)
 * - AnnotationList (Required)
 * - Range (Optional)
 * - Layer (Optional)
 * - Image Content (Required)
 * - Other Content (Required)
 */
const RESOURCE_TYPE_MAP = {
  'sc:Collection': 'collection',
  'sc:Sequence': 'sequence',
  'sc:Manifest': 'manifest',
  'sc:Canvas': 'canvas',
  'sc:AnnotationList': 'annotationList',
  'sc:Annotation': 'annotation',
  'sc:Range': 'range',
  'sc:Layer': 'layer',
  'dctypes:Image': 'imageResource',
};
const resource = new schema.Union(
  {
    collection,
    sequence,
    manifest,
    canvas,
    annotationList,
    annotation,
    range,
    layer,
    imageResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity['@type']]) {
      return RESOURCE_TYPE_MAP[entity['@type']];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  collections: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity['@type'] === 'sc:Manifest' ? 'manifest' : 'collection')
  ),
  manifests: [manifest],
  members: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity['@type'] === 'sc:Manifest' ? 'manifest' : 'collection')
  ),

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  sequences: [sequence],
  structures: [range],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 3) Sequences
// ===========================================================================
sequence.define({
  canvases: [canvas],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,
  startCanvas: canvas,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 4) Canvases
// ===========================================================================
canvas.define({
  images: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  otherContent: [annotationList],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 5) Annotation
// ===========================================================================
annotation.define({
  resource: new schema.Union(
    {
      imageResource,
      externalResource,
      choice,
    },
    input => {
      switch (input['@type']) {
        case 'dctypes:Image':
          return 'imageResource';
        case 'oa:Choice':
          return 'choice';
        default:
          return 'externalResource';
      }
    }
  ),
  on: canvas,

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 6) Annotation List
// ===========================================================================
annotationList.define({
  resources: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  resources: [annotation],
  members: [rangeMember],
  canvases: [canvas],
  ranges: [range],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,
  startCanvas: canvas,
  contentLayer: layer,

  // Extra
  thumbnail: imageResource,
});

// ===========================================================================
// 8) Layer
// ===========================================================================
layer.define({
  // Linking
  seeAlso: [externalResource],
  service: [service],
  related: [externalResource],
  rendering: [externalResource],
  within: within,
});

// ===========================================================================
// 9) Image resource
// ===========================================================================
imageResource.define({
  service: [service],
});

// ===========================================================================
// 10) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

const preprocess = compose(
  moveStartCanvasToSequence,
  preprocessLinkedEntities,
  addMissingIds
);

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  normalizeResource as normalize,
  preprocess,
  resource,
  collection,
  manifest,
  sequence,
  canvas,
  annotation,
  annotationList,
  range,
  layer,
  imageResource,
  externalResource,
  within,
  service,
};
