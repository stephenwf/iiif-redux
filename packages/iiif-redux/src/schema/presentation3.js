import { schema, normalize } from 'normalizr';
import { compose } from 'redux';
import addMissingIds from '../compat/addMissingIds';
import parseSelectorTarget from '../utility/parseTargetSelector';

function createEntity(name, processStrategy) {
  const options = {
    idAttribute: 'id',
  };
  if (processStrategy) {
    options.processStrategy = processStrategy;
  }

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections'); // 1
const manifest = createEntity('manifests'); // 2
const canvas = createEntity('canvases'); // 3
const annotation = createEntity('annotations', original => {
  const body = Array.isArray(original.body) ? original.body : [original.body];
  const target = Array.isArray(original.target)
    ? original.target
    : [original.target];
  const newObj = { ...original };
  if (original.body) {
    newObj.body = body;
  }
  if (original.target) {
    newObj.target = target;
  }
  return newObj;
}); // 4
const annotationPage = createEntity('annotationPages'); // 5
const annotationCollection = createEntity('annotationCollections'); // 6
const range = createEntity('ranges'); // 7
const contentResource = createEntity('contentResources'); // 8
const choice = createEntity('choices'); // 9
const canvasReference = createEntity('canvasReferences'); // 10
const selector = new schema.Entity(
  'selectors',
  {
    id: canvas,
  },
  {
    processStrategy(value, parent, key) {
      // @todo This is an impossible path to hit. The process strategy is a strange function that
      // doesn't always hit the containing code. Will need to review at a later date.
      // if (typeof value === 'string') {
      //   return parseSelectorTarget(value);
      // }
      const id = value['@id'] || value.id;
      if (id) {
        return parseSelectorTarget(id);
      }
      return value;
    },
  }
);

// Unofficial types.
const service = createEntity('services');

// Union types
const partOf = new schema.Array(
  {
    collection,
    manifest,
    contentResource,
    annotationCollection,
  },
  entity => {
    switch (entity.type) {
      case 'Collection':
        return 'collection';
      case 'Manifest':
        return 'manifest';
      case 'AnnotationCollection':
        return 'annotationCollection';
      default:
        return 'contentResource';
    }
  }
);

const canvasOrReference = new schema.Union(
  {
    selector,
    canvasReference,
    canvas,
  },
  entity => {
    if (entity && entity.id && entity.id.indexOf('#') !== -1) {
      return 'selector';
    }
    if (entity.type === 'SpecificResource') {
      return 'canvasReference';
    }
    return 'canvas';
  }
);

const annotationBodyMappings = {
  Choice: 'choice',
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const annotationBody = new schema.Union(
  {
    contentResource,
    choice,
  },
  input => {
    if (annotationBodyMappings[input.type]) {
      return annotationBodyMappings[input.type];
    }
    return 'contentResource';
  }
);

const rangeItem = new schema.Union(
  {
    canvas,
    selector,
    range,
    canvasReference,
  },
  input => {
    if (input && input.id && input.id.indexOf('#') !== -1) {
      return 'selector';
    }

    switch (input.type) {
      case 'Canvas':
        return 'canvas';
      case 'Range':
        return 'range';
      case 'SpecificResource':
      default:
        // @todo default?
        return 'canvasReference';
    }
  }
);

// Root resource
/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - Canvas (Recommended)
 * - Annotation (Recommended)
 * - AnnotationPage (Required)
 * - AnnotationCollection (Optional)
 * - Content Resource (Required)
 * - Range (Optional)
 */
const RESOURCE_TYPE_MAP = {
  Collection: 'collection',
  Sequence: 'sequence',
  Manifest: 'manifest',
  Canvas: 'canvas',
  AnnotationCollection: 'annotationCollection',
  AnnotationPage: 'annotationPage',
  Annotation: 'annotation',
  Range: 'range',
  // Content resources.
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const resource = new schema.Union(
  {
    collection,
    manifest,
    canvas,
    annotationCollection,
    annotationPage,
    annotation,
    range,
    contentResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity.type]) {
      return RESOURCE_TYPE_MAP[entity.type];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  // Structural
  items: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity.type === 'Manifest' ? 'manifest' : 'collection')
  ),
  annotations: [annotationPage],

  // Linking.
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra.
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  items: [canvas],
  structures: [range],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,
  start: canvasOrReference,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 3) Canvases
// ===========================================================================
canvas.define({
  // Structural.
  items: [annotationPage],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 4) Annotation
// ===========================================================================
annotation.define({
  // Structural.
  body: [annotationBody],
  target: [selector],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 5) Annotation Collection
// ===========================================================================
annotationCollection.define({
  // Structural.
  items: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // extra
  thumbnail: [contentResource],

  // Paging.
  first: annotationPage,
  last: annotationPage,
});

// ===========================================================================
// 6) Annotation Page
// ===========================================================================
annotationPage.define({
  // Structural.
  items: [annotation],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],

  // Paging.
  next: annotationPage,
  previous: annotationPage,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  // Structural.
  items: [rangeItem],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,
  start: canvasOrReference,
  supplementary: annotationCollection,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 8) Content Resource
// ===========================================================================
contentResource.define({
  // Linking.
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  rendering: [contentResource],
  partOf: partOf,
  // Structural
  annotations: [annotationPage],
});

// ===========================================================================
// 9) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

// ===========================================================================
// 10) Canvas Reference
// ===========================================================================
canvasReference.define({
  source: canvas,
});

const preprocess = compose(addMissingIds(3.0));

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  collection,
  manifest,
  canvas,
  annotation,
  annotationPage,
  annotationCollection,
  range,
  contentResource,
  choice,
  canvasReference,
  service,
  partOf,
  annotationBody,
  rangeItem,
  RESOURCE_TYPE_MAP,
  resource,
  preprocess,
  normalizeResource as normalize,
};
