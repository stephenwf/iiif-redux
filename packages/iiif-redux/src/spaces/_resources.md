# Resources
Different resources will track different state (only their own state.)

This is a reference implementation of how state could be represented and could be
thrown out and extended in a custom Redux implementation.

## Collection state
```js
const state = {
  currentCollection: 'http://.../',
}
```

## Manifest state
```js
const state = {
  currentSequence: 'http://.../',
  currentRange: 'http://.../',
}
```

## Sequence
```js
const state = {
  currentCanvas: 'http://.../',
}
```

## Canvas
```js
const state = {
  visibleOtherContentLists: [ /* ... */ ],
  thumbnailSize: null,
}
```

