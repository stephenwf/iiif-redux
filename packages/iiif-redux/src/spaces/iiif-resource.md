# IIIF Resources
This state is where all of our entities get imported into. The actions here are dynamic
and will call off to other spaces with follow up actions.

## State structure
```js
const state = {
  dereferenced: {
    'http://.../collection.json': { 
      loading: true,
      requested: '00-00-0000:00:00:00', 
      ttl: 600,
      error: null, // or { message: 'Failed to fetch', ... }
    },
  },
  collections: {
    'http://.../collection.json': {/* ... */},
  },
  manifests: {},
  sequences: {},
  canvass: {},
  annotations: {},
  annotationLists: {},
  ranges: {},
  layers: {},
  imageResources: {},
  externalResources: {},
  services: {},
}
```

## Reusable actions
Requesting an item that can definitely be referenced (i.e. manifest):
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    schema: schemas.manifest,
  }
});
```

Requesting an items fields, dereferencing if they don't exist.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    dereferenceIfMissing: ['sequences'],
    schema: schemas.imageService,
  }
});
```

Requesting a resource, with a TTL so if its requested again a fresh copy may be requested.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    schema: schemas.manifest,
  }
});
```

Force-requesting an item, includes cache busting
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    forceFresh: true,
    schema: schemas.manifest,
  }
});
```

## Helper selectors
For use in containers for a more verbose way of showing resources current state while requesting.
```js
isResourceLoaded(id); // returns true/false, for loading screens.
didResourceError(id); // returns error object or null
```

## Potential sagas
In each of the spaces for the different resource types, we can have different reactions and
store different state. 

```js
function* manifestSaga() {
  yield takeEvery('MANIFEST_SUCCESS', handleSuccess);
} 

function* handleSuccess({ payload: { id } }) {
  // Set the current manifest to the one we just requested.
  yield put(setCurrentManifest(id));
  
  // Grab the state + start canvas
  const state = yield select();
  const startCanvas = getStartCanvas(state);

  // Set the canvas if its set.
  if (startCanvas) {
    yield put(setCurrentCanvasIndex(startCanvas)); // Another action creator.
  }
}
```
