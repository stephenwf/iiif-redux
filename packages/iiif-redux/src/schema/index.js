import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
