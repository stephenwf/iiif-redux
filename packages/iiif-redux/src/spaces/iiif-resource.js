import cachedFetch from 'fetch-unless-cached';
import { createActions, handleActions } from 'redux-actions';
import validUrl from 'valid-url';
import { call, put, select, takeEvery, all } from 'redux-saga/effects';
import deepmerge from 'deepmerge';
import update from 'immutability-helper';
import * as presentation2 from '../schema/presentation2';
import * as presentation3 from '../schema/presentation3';
import trackPresentationVersion from '../utility/trackPresentationVersion';
import {
  COLLECTION_ERROR,
  COLLECTION_REQUEST,
  COLLECTION_SUCCESS,
} from './collections';
import {
  MANIFEST_ERROR,
  MANIFEST_REQUEST,
  MANIFEST_SUCCESS,
} from './manifests';
import getPresentationVersionFromResource from '../utility/getPresentationVersionFromResource';
const debug = require('debug')('iiif-redux');

const IIIF_RESOURCE_REQUEST = 'IIIF_RESOURCE_REQUEST';
const IIIF_RESOURCE_SUCCESS = 'IIIF_RESOURCE_SUCCESS';
const IIIF_RESOURCE_ERROR = 'IIIF_RESOURCE_ERROR';
const IIIF_RESOURCE_REQUEST_UNKNOWN = 'IIIF_RESOURCE_REQUEST_UNKNOWN';
const IIIF_RESOURCE_IDENTIFY = 'IIIF_RESOURCE_IDENTIFY';

const DEFAULT_MAPPINGS = {
  collection: [COLLECTION_REQUEST, COLLECTION_SUCCESS, COLLECTION_ERROR],
  manifest: [MANIFEST_REQUEST, MANIFEST_SUCCESS, MANIFEST_ERROR],
};

const {
  iiifResourceRequestUnknown,
  iiifResourceRequest,
  iiifResourceSuccess,
  iiifResourceError,
} = createActions({
  [IIIF_RESOURCE_REQUEST_UNKNOWN]: (
    resourceId,
    options = {},
    mappings = DEFAULT_MAPPINGS
  ) => ({
    resourceId,
    dereferenceIfMissing: options.dereferenceIfMissing || false,
    ttl: options.ttl || 600,
    forceFresh: options.forceFresh,
    mappings,
  }),
  [IIIF_RESOURCE_REQUEST]: (resourceId, types, schema, options = {}) => ({
    resourceId,
    types,
    schema,
    dereferenceIfMissing: options.dereferenceIfMissing || false,
    ttl: options.ttl || 600,
    forceFresh: options.forceFresh,
  }),
  [IIIF_RESOURCE_SUCCESS]: (resourceId, normalizedResponse) => ({
    resourceId,
    normalizedResponse,
  }),
  [IIIF_RESOURCE_ERROR]: (resourceId, error) => ({
    resourceId,
    error,
  }),
});

const DEFAULT_STATE = {
  dereferenced: {},
  resources: {
    collections: {},
    sequences: {},
    manifests: {},
    canvases: {},
    canvasReferences: {},
    annotations: {},
    annotationLists: {},
    annotationPages: {},
    annotationCollections: {},
    ranges: {},
    layers: {},
    imageResources: {},
    externalResources: {},
    contentResources: {},
    choices: {},
  },
};

const resourceReducer = handleActions(
  {
    [iiifResourceSuccess]: (state, { payload: { normalizedResponse } }) =>
      deepmerge(state, normalizedResponse),
  },
  DEFAULT_STATE.resources
);

const dereferencedReducer = handleActions(
  {
    [iiifResourceRequest]: (state, { payload: { resourceId, ttl } }) =>
      update(state, {
        [resourceId]: {
          $set: {
            resourceId,
            ttl,
            requested: state[resourceId]
              ? state[resourceId].requested
              : new Date(),
            loading: state[resourceId] ? state[resourceId].loading : true,
          },
        },
      }),
    [iiifResourceRequestUnknown]: (state, { payload: { resourceId, ttl } }) =>
      update(state, {
        [resourceId]: {
          $set: {
            resourceId,
            ttl,
            loading: state[resourceId] ? state[resourceId].loading : true,
          },
        },
      }),

    [iiifResourceError]: (state, { payload: { resourceId, error } }) =>
      update(state, {
        [resourceId]: {
          loading: { $set: false },
          error: { $set: error },
        },
      }),

    [iiifResourceSuccess]: (state, { payload: { resourceId } }) =>
      update(state, {
        [resourceId]: {
          loading: { $set: false },
        },
      }),
  },
  DEFAULT_STATE.dereferenced
);

async function requestResource(resourceId, options) {
  /* istanbul ignore if */
  if (process.env.NODE_ENV === 'production') {
    return cachedFetch(resourceId);
  }

  return fetch(resourceId).then(resp => resp.json());
}

function* errorAction(type, resourceId, error) {
  yield put(iiifResourceError(resourceId, error));
  if (type) {
    yield put({ type, payload: { resourceId, error } });
  }
}

function* successAction(type, resourceId, normalizedResponse) {
  yield put(iiifResourceSuccess(resourceId, normalizedResponse));
  yield put({ type, payload: { resourceId, normalizedResponse } });
}

function* requestIiifResource({ payload }) {
  const { resourceId, types, ...options } = payload;
  if (types.length !== 3 && process.env.NODE_ENV !== 'production') {
    yield call(
      errorAction,
      null,
      resourceId,
      'DEV ERROR: You must pass in exactly 3 action types'
    );
  }

  const [REQUEST, SUCCESS, ERROR] = types;

  if (!validUrl.isWebUri(resourceId)) {
    yield call(
      errorAction,
      ERROR,
      resourceId,
      'Resource is not a valid Web URI.'
    );
    return;
  }

  const state = yield select();

  // @todo logic for cache.
  if (
    state.dereferenced[resourceId] &&
    state.dereferenced[resourceId].loading ===
      false /*&& options.forceFresh === false*/
  ) {
    debug('Skipping fetch for resource %s using cache.', resourceId);

    return;
  }

  yield put({ type: REQUEST, payload });

  try {
    debug('Fetching resource %s', resourceId);
    const response = yield call(requestResource, resourceId, options);

    const { result, entities } = yield call(
      importResource,
      response,
      resourceId
    );
    debug('Finished normalize resource %s', result.id);

    yield call(successAction, SUCCESS, result.id, entities);
  } catch (err) {
    debug('Error: %O', err);
    yield call(errorAction, ERROR, resourceId, err);
  }
}

function* importResource(response, resourceId) {
  const version = getPresentationVersionFromResource(response);
  if (version === 2) {
    return yield call(importPresentation2Resource, response, resourceId);
  }
  if (version === 3) {
    return yield call(importPresentation3Resource, response, resourceId);
  }

  throw new Error('Unknown IIIF Presentation version');
}

function* importPresentation2Resource(response, resourceId) {
  if (response['@id'] !== resourceId) {
    // @todo this will not catch "partOf" fields.
    debug(
      'Resource ID does not match requested presentation 2 resource, patching... Found: %s Expected: %s',
      response['@id'],
      resourceId
    );
    response['@id'] = resourceId;
  }

  debug('Starting normalize presentation 2 resource %s', resourceId);

  return trackPresentationVersion(
    2,
    presentation2.normalize(response, presentation2.resource)
  );
}

function* importPresentation3Resource(response, resourceId) {
  if (response.id !== resourceId) {
    // @todo this will not catch "partOf" fields.
    debug(
      'Resource ID does not match requested presentation 3 resource, patching... Found: %s Expected: %s',
      response.id,
      resourceId
    );
    response.id = resourceId;
  }

  debug('Starting normalize presentation 3 resource %s', resourceId);

  return trackPresentationVersion(
    3,
    presentation3.normalize(response, presentation3.resource)
  );
}

function* getResourceType(resourceId, response) {
  const presentationVersion = getPresentationVersionFromResource(response);
  if (presentationVersion === 2) {
    return yield call(getPresentation2ResourceType, resourceId, response);
  }
  if (presentationVersion === 3) {
    return yield call(getPresentation3ResourceType, resourceId, response);
  }

  throw new Error('Unknown IIIF Presentation version');
}

function* getPresentation2ResourceType(resourceId, response) {
  if (!presentation2.RESOURCE_TYPE_MAP[response['@type']]) {
    throw new Error('Unknown Presentation 2 resource type');
  }

  return presentation2.RESOURCE_TYPE_MAP[response['@type']];
}

function* getPresentation3ResourceType(resourceId, response) {
  if (!presentation3.RESOURCE_TYPE_MAP[response.type]) {
    throw new Error('Unknown Presentation 3 resource type');
  }

  return presentation3.RESOURCE_TYPE_MAP[response.type];
}

function* requestUnknownResource({
  payload: { resourceId, mappings, ...options },
}) {
  if (!validUrl.isWebUri(resourceId)) {
    yield put(iiifResourceError(resourceId, 'Resource is not a valid URL.'));
    return;
  }

  try {
    // 1) Request the resource.
    debug('Fetching unknown resource %s', resourceId);
    const response = yield call(requestResource, resourceId, options);
    if (!response) {
      yield put(iiifResourceError(resourceId, 'Unknown resource'));
      return;
    }

    const type = yield call(getResourceType, resourceId, response);

    if (!mappings[type]) {
      yield put(
        iiifResourceError(
          resourceId,
          `Resource type is not in configured mappings (${Object.keys(
            mappings
          ).join(', ')})`
        )
      );
      return;
    }

    // 3) Grab constants from a map (to be created).
    const [REQUEST, SUCCESS, ERROR] = mappings[type];

    // 4) Send to regular import (will be cached).
    yield put(iiifResourceRequest(resourceId, [REQUEST, SUCCESS, ERROR]));
  } catch (err) {
    yield put(iiifResourceError(resourceId, err.message));
  }
}

function* saga() {
  yield all([
    takeEvery(IIIF_RESOURCE_REQUEST, requestIiifResource),
    takeEvery(IIIF_RESOURCE_REQUEST_UNKNOWN, requestUnknownResource),
  ]);
}

export {
  saga,
  resourceReducer,
  dereferencedReducer,
  iiifResourceRequest,
  iiifResourceRequestUnknown,
  DEFAULT_STATE,
  IIIF_RESOURCE_REQUEST,
  IIIF_RESOURCE_SUCCESS,
  IIIF_RESOURCE_ERROR,
  IIIF_RESOURCE_REQUEST_UNKNOWN,
  IIIF_RESOURCE_IDENTIFY,
};
