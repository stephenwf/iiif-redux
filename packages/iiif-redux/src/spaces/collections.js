/**
 * Collections space
 *
 * The logic for collections.
 *
 * What is known:
 * - Collections can contain manifests
 * - Collections can contain other collections
 *
 * Interactions:
 * - When collection is selected, it should be loaded if its not already.
 *
 * Needs planning:
 * - Next/Prev
 */
import { put, all, takeEvery } from 'redux-saga/effects';
import { ROUTING_SELECT_COLLECTION } from './routing';
import { collection } from '../schema/presentation2';
import { iiifResourceRequest } from './iiif-resource';

const COLLECTION_REQUEST = 'COLLECTION_REQUEST';
const COLLECTION_SUCCESS = 'COLLECTION_SUCCESS';
const COLLECTION_ERROR = 'COLLECTION_ERROR';

function* fetchCollection({ payload: { id } }) {
  yield put(
    iiifResourceRequest(
      id,
      [COLLECTION_REQUEST, COLLECTION_SUCCESS, COLLECTION_ERROR],
      collection
    )
  );
}

function* saga() {
  yield all([takeEvery(ROUTING_SELECT_COLLECTION, fetchCollection)]);
}

export { COLLECTION_SUCCESS, COLLECTION_REQUEST, COLLECTION_ERROR, saga };
