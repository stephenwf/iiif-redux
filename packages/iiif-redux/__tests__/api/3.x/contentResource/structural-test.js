import contentResource from '../../../../src/api/3.x/content-resource';
import { normalize } from '../../../../src/schema/presentation3';
import fixtures from './fixtures';
import { DEFAULT_STATE } from '../../../../src/spaces/iiif-resource';

describe('api/3.x/content-resource/structural', () => {
  describe.each(fixtures)('Fixture: %s', (name, fixture) => {
    const {
      result: { id },
      entities,
    } = normalize(fixture);
    const state = {
      resources: { ...DEFAULT_STATE.resources, ...entities },
    };
    const selectContentResource = s => s.resources.contentResources[id];

    const { getAnnotationIds, getAnnotations } = contentResource(
      selectContentResource
    );

    test('getAnnotationIds', () => {
      expect(getAnnotationIds(state)).toMatchSnapshot('getAnnotationIds');
    });
    test('getAnnotations', () => {
      expect(getAnnotations(state)).toMatchSnapshot('getAnnotations');
    });
  });
});
