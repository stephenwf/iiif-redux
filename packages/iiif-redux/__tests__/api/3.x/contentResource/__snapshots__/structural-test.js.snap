// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/3.x/content-resource/structural Fixture: Full content resource getAnnotationIds: getAnnotationIds 1`] = `
Array [
  "https://example.org/iiif/book1/page/content-resource/1",
]
`;

exports[`api/3.x/content-resource/structural Fixture: Full content resource getAnnotations: getAnnotations 1`] = `
Array [
  Object {
    "id": "https://example.org/iiif/book1/page/content-resource/1",
    "items": Array [
      "https://example.org/iiif/book1/content-resource/a1",
    ],
    "type": "AnnotationPage",
  },
]
`;

exports[`api/3.x/content-resource/structural Fixture: IIIF spec example getAnnotationIds: getAnnotationIds 1`] = `Array []`;

exports[`api/3.x/content-resource/structural Fixture: IIIF spec example getAnnotations: getAnnotations 1`] = `Array []`;

exports[`api/3.x/content-resource/structural Fixture: Minimum content resource getAnnotationIds: getAnnotationIds 1`] = `Array []`;

exports[`api/3.x/content-resource/structural Fixture: Minimum content resource getAnnotations: getAnnotations 1`] = `Array []`;

exports[`api/3.x/content-resource/structural Fixture: Thumbnail getAnnotationIds: getAnnotationIds 1`] = `Array []`;

exports[`api/3.x/content-resource/structural Fixture: Thumbnail getAnnotations: getAnnotations 1`] = `Array []`;
